service/raft/raft_group0.cc

```diff
@@ -110,6 +110,13 @@ raft_server_for_group raft_group0::create_server_for_group(raft::group_id gid,
     };
 }
 
+static future<group0_info> run_discovery(
```

Comment:
Nice. It would be great to make the name explain why we had to out-of-line this code. Maybe make the tracker a public class and make the method part of the tracker? Technically, the main goal of having it inside a separate function is to be able to "disconnect" its life cycle from the life cycle of group0.

Reply:
Note that `tracker` is specific to a discovery loop iteration. Each iteration gets a new `tracker`. So it wouldn't make much sense to make it a member of `tracker` (unless you mean a static method?)

Maybe it would make sense to make it a member of `persistent_discovery`.

In theory `run_discovery` could become completely independent of the group 0 context - it's more generic, it's just running the discovery algorithm (generally, the discovery algorithm is just for electing a leader in a set of nodes that are connected by a tree of dependencies - it doesn't necessarily have anything to do with raft groups). Unfortunately, it's still calling "group 0 specific" functions such as `send_group0_peer_exchange`. These could be abstracted out (e.g. the send function could be passed as a parameter). But I didn't want to go that far, so I left it in this half-refactored state.